src/net/netsyscall.c

```diff
@@ -641,7 +670,8 @@ closure_function(6, 1, sysreturn, socket_write_tcp_bh,
         goto out_unlock;
     }
 
-    struct tcp_pcb *tcp_lw = s->info.tcp.lw;
+    context ctx = get_current_context(current_cpu());
+    struct tcp_pcb * volatile tcp_lw = s->info.tcp.lw;
```

Comment:
It looks as if the volatile is used here to prevent fields of the struct from being cached in registers across an error frame apply, as you state in the description. I'm wondering how one can definitively know whether or not to use volatile. The possibility of missing a critical volatile could be a source of risk in introducing the error frame mechanism.

I suggest that a more sound approach may be to call compiler_barrier() within context_set_err() prior to the error frame save. If I recall the semantics of the barrier correctly, it should require the compiler to not cache memory content across the barrier. If so, this may obviate the need for these volatiles.

I suppose this could also mean that some stores to memory may not be committed from a cached register prior to a fault, but then the side effects of such transient state should be considered independently - and perhaps ameliorated with additional, judiciously-placed barriers prior to potentially faulting code.

Reply:
gcc apparently has a special "returns_twice" function attribute intended for use with such longjmp-like functions:

> The returns_twice attribute tells the compiler that a function may return more than one time. The compiler ensures that all registers are dead before calling such a function and emits a warning about the variables that may be clobbered after the second return from the function. Examples of such functions are setjmp and vfork. The longjmp-like counterpart of such function, if any, might need to be marked with the noreturn attribute.

https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-returns_005ftwice-function-attribute